[정승호] 회원가입/로그인 엔드포인트 (feature/user)

seungho-jeong opened the pull request:
## 수정 사항 간략한 한줄 요약
#### Commit 1
- 회원가입 및 로그인 엔드포인트 생성
#### Commit 2
- 불필요한 문장 삭제, Status code 및 에러 처리방식 변경 등
#### Commit 3
- 언어 스타일에 맞지 않는 문 삭제, Login 에러 메시지 보완
## 수정 사항들 자세한 내용
#### Commit 1
- 회원가입 시 패스워드 Hashing 적용
- Response 시 JWT Token 생성 및 반환되도록 적용
- 회원가입/로그인 시 예상되는 에러 예외처리 적용
#### Commit 2
- 배포 전 테스트 목적으로 작성한 코드 삭제
- 자원 생성 결과 메시지 Stat. code를 200에서 201로 변경
- 권한 관련 에러 메시지 Stat. code를 400에서 401로 변경
- Get method로 Object를 조회하는 경우 발생하는 에러 캐치를 if statement에서 Model.DoesNotExist 방식으로 변경
- Request body 중 형식을 유지해야 하는 값(이메일, 전화번호)은 포맷을 식별하도록 정규식 적용
#### Commit 3
- SignupView 내용 중 프로그램 동작상 존재유무가 관계없는 Else 문 삭제
- LoginView에서 User or PW 중 잘못된 값이 전달되었을 때 발생하는 에러 메시지를 동일하게 보완

## 기타 질문 및 특이 사항

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

seungho-jeong committed:
Add: 회원가입/로그인 엔드포인트 생성
- 회원가입 시 패스워드 Hashing 적용
- Response 시 JWT Token 생성 및 반환
- 회원가입/로그인 시 예상되는 에러 예외처리 적용

lordmyshepherd commented on user/views.py:
```diff
@@ -0,0 +1,103 @@
+class SignUpView(View):  # 클래스명 협의 필요
+    def post(self, request):
+        data = json.loads(request.body)
+        print("제이슨 로즈 성공")
```
print 문은 지워주세요~

Dan-Jeong replied:
삭제하였습니다

seungho-jeong replied:
:white_check_mark: 삭제하였습니다

lordmyshepherd commented on user/views.py:
```diff
@@ -0,0 +1,103 @@
+                    mobile=data["mobile"],
+                    password=hashed_password.decode("UTF-8")
+                )
+                return JsonResponse({"message": "SUCCESS"}, status=200)
```
자원(resource)을 생성할 경우 status code는 201이 더 정확합니다~

seungho-jeong replied:
반영하였습니다.
추가로 더 찾아보니 권한 관련 Error(Unauthrization)의 경우 401 적용하는 것이 의미상 맞는 것 같아 수정하였습니다 :slightly_smiling_face:

lordmyshepherd commented on user/views.py:
```diff
@@ -0,0 +1,103 @@
+        except KeyError as e:
+            return JsonResponse({"message": f"KEY_ERROR:{e}"}, status=400)
+
+        except ValueError as e:
```
valueError는 언제 발생하게 되나요??

seungho-jeong replied:
불필요한 예외 처리같아 삭제하였습니다.

lordmyshepherd commented on user/views.py:
```diff
@@ -0,0 +1,103 @@
+        except ValueError as e:
+            return JsonResponse({"message": f"VALUE_ERROR:{e}"}, status=400)
+
+        except:
+            return JsonResponse({"message": "INVALID_ERROR"}, status=400)
```
```suggestion
```

seungho-jeong replied:
로그인 시 발생할 수 있는 에러는 KeyError에 국한되는 것 같아 해당 에러는 삭제하였습니다.

lordmyshepherd commented on user/views.py:
```diff
@@ -0,0 +1,103 @@
+        except ValueError as e:
+            return JsonResponse({"message": f"VALUE_ERROR:{e}"}, status=400)
+
+        except:
```
예측할 수 없는 에러가 발생할 경우, INVALID_ERROR를 리턴해 주는 것 보다는 에러를 raise를 하는 편이 더 좋습니다.

seungho-jeong replied:
로그인 시 발생할 수 있는 에러는 KeyError에 국한되는 것 같아 해당 에러는 삭제하였습니다.

lordmyshepherd commented on user/views.py:
```diff
@@ -0,0 +1,103 @@
+                    mobile=data["mobile"],
+                    password=hashed_password.decode("UTF-8")
+                )
+                return JsonResponse({"message": "SUCCESS"}, status=200)
```
자원을 생성하는 경우 `status 201`을 사용해주세요.

seungho-jeong replied:
코멘트 주신대로 201코드로 변경하였고 추후 코멘트 등 자원 생성 관련 코드에도 적용하겠습니다.
반환 메시지도 보다 명확하게 Registration success로 수정하였습니다.

lordmyshepherd commented on user/views.py:
```diff
@@ -0,0 +1,103 @@
+        except:
+            return JsonResponse({"message": "INVALID_ERROR"}, status=400)
+
+    def get(self, request):
```
해당 함수의 역할은 무었인가요??
모든 users를 리턴받는 경우가 있나요??

seungho-jeong replied:
여러 테스트 시 작성하였던 코드로 배포상 불필요하여 삭제하였습니다.

lordmyshepherd commented on user/views.py:
```diff
@@ -0,0 +1,103 @@
+class LoginView(View):
+    def post(self, request):
+
+        try:
+            data = json.loads(request.body)
+            given_email = data["email"]
+            given_pw = data["password"]
+
+            if User.objects.filter(email=given_email).exists():
+                account = User.objects.get(email=given_email)
+
+                if account:
+
+                    if bcrypt.checkpw(given_pw.encode("UTF-8"), account.password.encode("UTF-8")):
+                        token = jwt.encode(
+                            {"email": given_email},
+                            my_settings.SECRET["secret"],
+                            algorithm=my_settings.ALGORITHM["algorithm"]
+                        )
+                        token = token.decode("UTF-8")
+                        return JsonResponse({"message": "SUCCESS", "token": token}, status=200)
+
+                    else:
+                        return JsonResponse({"message": "WRONG_PASSWORD"}, status=401)
+
+            else:
+                return JsonResponse({"message": "DOES_NOT_EXIST_USER"}, status=401)
+
+        except KeyError as e:
+            return JsonResponse({"message": f"KEY_ERROR:{e}"}, status=400)
+
```
```suggestion
        try:
            data = json.loads(request.body)
            if User.objects.filter(email=data['email']l).exists():
                user = User.objects.get(email=data['email'])
                if bcrypt.checkpw(given_pw.encode("UTF-8"), account.password.encode("UTF-8")):
                    token = jwt.encode(
                        {"email": given_email},
                        my_settings.SECRET["secret"],
                        algorithm=my_settings.ALGORITHM["algorithm"]
                        )
                        token = token.decode("UTF-8")
                        return JsonResponse({"message": "SUCCESS", "token": token}, status=200)

                    else:
                        return JsonResponse({"message": "WRONG_PASSWORD"}, status=401)

            else:
                return JsonResponse({"message": "DOES_NOT_EXIST_USER"}, status=401)

        except KeyError as e:
            return JsonResponse({"message": f"KEY_ERROR:{e}"}, status=400)

```

lordmyshepherd commented on user/views.py:
```diff
@@ -0,0 +1,103 @@
+            if User.objects.filter(email=given_email).exists():
+                account = User.objects.get(email=given_email)
+
+                if account:
```
User.objects.get()으로 조회하는 경우, 존재하지 않을시 DoesNotExist를 리턴해주기 때문에 if account:가 의미 없어보입니다.
또한 get()을 사용하셨으면 User.DoesNotExist 에러를 케치해주셔야 합니다!

seungho-jeong replied:
장고 공식문서를 찾아보니 관련 설명이 잘되어 있네요. 몰랐던 부분인데 잘 알려주셔서 감사합니다 :slightly_smiling_face:
if 문을 삭제하고 DoesNotExist 에러 반영하였습니다

seungho-jeong committed:
Modify: 불필요 문장 삭제, Status code 변경 및 에러 처리방식 변경 등
- 배포 전 테스트 목적으로 작성한 코드 삭제
- 자원 생성 결과 메시지 Status code를 200에서 201로 변경
- 권한 관련 에러 메시지 Status code를 400에서 401로 변경
- Get method로 OBJ를 조회하는 경우 발생하는 에러 캐치를
  if statement 방식에서  Model.DoesNotExist형으로 변경
- Request body 중 형식을 유지해야 하는 값(email, mobile)은 포맷을 식별토록 정규식 적용

soheon-lee commented on user/views.py:
```diff
@@ -0,0 +1,75 @@
+            if not re.match("^[a-zA-Z0-9]([-_\.]?[a-zA-Z0-9])*@[a-zA-Z0-9]([-_\.]?[a-zA-Z0-9])*\.[a-zA-Z]{2,}$", data["email"]):
+                return JsonResponse({"MESSAGE": "NOT_IN_EMAIL_FORM"}, status=400)
+
+            else:
+                hashed_password = bcrypt.hashpw(
+                    data["password"].encode("UTF-8"), bcrypt.gensalt())
+                User.objects.create(
+                    user_name         = data["user_name"],
+                    email             = data["email"],
+                    mobile            = data["mobile"],
+                    password          = hashed_password.decode("UTF-8"),
+                    introduction      = data["introduction"],
+                    profile_image_url = data["profile_image_url"]
+                )
+                return JsonResponse({"MESSAGE": "REGISTRATION_SUCCESS"}, status=201)
```
else가 없어도 바로 위의 if문이 만족되지 않는다면 바로 도달할 수 있는 로직입니다. 꼭 필요한 경우가 아니면 else를 사용하지 않는 것이 보다 파이썬다운 코딩 스타일입니다. 지워주셔도 됩니다 :)

seungho-jeong replied:
IF문이 있으면 Else가 한 군데는 꼭 있어야 한다는 편견을 갖고 있었는데,
동작상 문제가 없으면 생략하는 것이 보다 Pythonic한 방식이군요!

코멘트주신 부분 반영하였습니다. 잘 알려주셔서 감사합니다 :slightly_smiling_face:

soheon-lee commented on user/views.py:
```diff
@@ -0,0 +1,75 @@
+                )
+                return JsonResponse({"MESSAGE": "LOGIN_SUCCESS", "token": token.decode("UTF-8")}, status=200)
+
+            else:
+                return JsonResponse({"MESSAGE": "WRONG_PASSWORD"}, status=401)
+
+        except User.DoesNotExist:
+                return JsonResponse({"MESSAGE": "DOES_NOT_EXIST_USER"}, status=401)
```
유저의 정보가 옳지 않을 때와, 비밀번호가 옳지 않을 때에 서로 다른 메세지를 전달하는 것은 보안상 취약합니다 :) 따라서 이메일과 비밀번호가 적절하지 않은 두 가지 모두 "INVALID_INPUT" 등과 같은 동일한 메세지를 전달하시는 것을 추천드립니다

seungho-jeong replied:
악의적인 의도로 계정 취득 시도시 예측범위를 줄일 수 있는 에러 메시지는 보완하였습니다.
다른 메시지로 적용하려고 했으나 INVALID_INPUT이 가장 적절하다고 생각하여 그대로 채택하였습니다.

seungho-jeong committed:
Modify: 언어 스타일에 맞지 않는 문 삭제, Login 에러 메시지 보완
- SignupView 내용 중 프로그램 동작상 존재유무가 상관없는 else문 삭제
- LoginView 유저 or PW 잘못된 값 전달 시 발생하는 에러 메시지 통일

soheon-lee commented:
LGTM
